File: solutions/kth-largest-element-in-an-array/kth-largest-element-in-an-array.cpp
// @Title: Kth Largest Element in an Array
// @Author: colinjxc
// @Date: 2022-02-07T01:42:31+08:00
// @URL: https://leetcode-cn.com/problems/kth-largest-element-in-an-array


class Solution {
private:
    int FindKthLargest_(vector<int> &nums, int k,  int l, int r) {
        if (l >= r) {
            return nums[l];
        }

        int p = Partition(nums, l, r);
        if (p < k) {
            return FindKthLargest_(nums, k, p + 1, r);
        } else if (p > k) {
            return FindKthLargest_(nums, k, l, p - 1);
        } else {
            return nums[p];
        }
    }
    
    int Partition(vector<int> &nums, int l, int r) {
        int ri = random() % (r - l) + l;
        std::swap(nums[l], nums[ri]);

        int pv = nums[l];
        int p = l;

        for (int i = l + 1; i <= r; i++) {
            if (nums[i] > pv) {
                std::swap(nums[++p], nums[i]);
            }
        }
        std::swap(nums[p], nums[l]);
        return p;
    }
    
public:
    int findKthLargest(vector<int> &nums, int k) {
        int r = FindKthLargest_(nums, k - 1, 0, nums.size() - 1);
        return r;
    }
};

File: solutions/kth-largest-element-in-an-array/kth-largest-element-in-an-array.py
// @Title: Kth Largest Element in an Array
// @Author: colinjxc
// @Date: 2021-11-02T23:54:50+08:00
// @URL: https://leetcode-cn.com/problems/kth-largest-element-in-an-array


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def partition(nums, l, r) -> int:
            xx = random.randint(l, r)
            nums[xx], nums[l] = nums[l], nums[xx]
 
            p = l + 1
            for index in range(l+1, r+1):
                # 降序排序，如果发现大的就交换
                if nums[index] > nums[l]:
                    nums[p], nums[index] = nums[index], nums[p]
                    p = p + 1
            # 这块要注意啊，是 p - 1
            nums[p - 1], nums[l] = nums[l], nums[p -1]
            return p-1

    
        def findKth(nums, l, r, k) -> int:
            if l == r:
                return nums[l]

            p = partition(nums, l, r)                
            if p > k :
                return findKth(nums, l, p-1, k)
            elif p < k :
                return findKth(nums, p+1, r, k)
            return nums[k]


        return findKth(nums, 0, len(nums)-1, k -1)

        

    
